### PR TITLE
Add custom parser to `DragValue` and `Slider`, with additional support for binary, octal, and hexadecimal numbers.

### PR DESCRIPTION
**Pull request by Adanos020**
Closes <https://github.com/emilk/egui/issues/1953>.

Adds a `custom_parser` method to `DragValue` and `Slider`, along with extra `binary_u64`, `octal_u64`, and `hexadecimal_u64` which overwrite the `custom_formatter` and `custom_parser`.

The purpose of `custom_parser` is to convert the text input from both widgets to a `f64`. A parser function returns an `Option<f64>` set to `Some` when parsing succeeds or `None` when it fails.

Additional methods for different number systems all take in a `min_width` parameter that specifies the minimum number of displayed digits. If the number is shorter than `min_width`, it will be prefixed with additional 0s to match it. `hexadecimal_u64` also takes in a `bool` which specifies if in the output text alphabetic digits should be uppercase or not.

Demo (apologies for quality):
untitled

Code for the demo:
```rust
ui.add(
    egui::Slider::new(&mut self.age, 0..=1023)
        .text("bin")
        .binary_u64(10),
);
ui.add(
    egui::Slider::new(&mut self.age, 0..=1023)
        .text("oct")
        .octal_u64(4),
);
ui.add(
    egui::Slider::new(&mut self.age, 0..=1023)
        .text("hex")
        .hexadecimal_u64(3, false),
);
ui.add(
    egui::Slider::new(&mut self.age, 0..=1023)
        .text("HEX")
        .hexadecimal_u64(3, true),
);
ui.add(
    egui::Slider::new(&mut self.age, 0..=((60 * 60 * 24) - 1))
        .text("time")
        .custom_formatter(|n, _| {
            let n = n as i32;
            let hours = n / (60 * 60);
            let mins = (n / 60) % 60;
            let secs = n % 60;
            format!("{hours:02}:{mins:02}:{secs:02}")
        })
        .custom_parser(|s| {
            let parts: Vec<&str> = s.split(':').collect();
            if parts.len() == 3 {
                parts[0]
                    .parse::<i32>()
                    .and_then(|h| {
                        parts[1].parse::<i32>().and_then(|m| {
                            parts[2]
                                .parse::<i32>()
                                .map(|s| ((h * 60 * 60) + (m * 60) + s) as f64)
                        })
                    })
                    .ok()
            } else {
                None
            }
        }),
);
```

### CODE REVIEW

**Comment by Mingun** [comment]
May be return a `Result<T, String>` where the `Err` variant will represent a text that can be shown to the user why his input is incorrect? For now egui can just use `.ok()` and not shown anything, but in the future it could.

**Comment by Adanos020** [comment]
> May be return a `Result<T, String>` where the `Err` variant will represent a text that can be shown to the user why his input is incorrect? For now egui can just use `.ok()` and not shown anything, but in the future it could.

If we do want this, I think `Result<T, Option<String>>` would be better as it'd give you the option not to have an error message.

File: crates/egui/src/widgets/drag_value.rs [hunk 1]
```diff
@@ -170,6 +198,170 @@ impl<'a> DragValue<'a> {
...
+                    return None;
+                }
+                let num_digits = s.len().min(64);
+                let mut n = 0;
+                for (i, digit) in s.chars().rev().take(num_digits).enumerate() {
+                    if digit != '0' && digit != '1' {
+                        return None;
+                    }
+                    n |= (digit as u64 - '0' as u64) << (i as u64);
+                }
```
Comment (emilk) [inline crates/egui/src/widgets/drag_value.rs:276]: Use https://doc.rust-lang.org/stable/std/primitive.i64.html#method.from_str_radix here and below - less code, and automatic support for signed integers!

**Review by emilk — CHANGES_REQUESTED** [review]
nice!

**Comment by emilk** [comment]
> May be return a `Result<T, String>` where the `Err` variant will represent a text that can be shown to the user why his input is incorrect? For now egui can just use `.ok()` and not shown anything, but in the future it could.

Let's wait with that until we actually implement showing the errors (if ever)

File: crates/egui/src/widgets/drag_value.rs (hunk 1 above)
Comment (Adanos020) [reply crates/egui/src/widgets/drag_value.rs:276]: I was thinking about using that but wasn't sure if we wanted another dependency. I'll add it now!

Comment (Adanos020) [reply crates/egui/src/widgets/drag_value.rs:276]: Oh cool, I don't know how I missed this!

Comment (Adanos020) [reply crates/egui/src/widgets/drag_value.rs:276]: @emilk According to [`Binary`]() documentation:

> For primitive signed integers ([i8](https://doc.rust-lang.org/stable/std/primitive.i8.html) to [i128](https://doc.rust-lang.org/stable/std/primitive.i128.html), and [isize](https://doc.rust-lang.org/stable/std/primitive.isize.html)), negative values are formatted as the two’s complement representation.

Are we happy with this?

Comment (Adanos020) [reply crates/egui/src/widgets/drag_value.rs]: I added a bool that lets you choose if you want the number to be displayed with a sign or in 2's complement.

**Comment by Adanos020** [comment]
Applied suggested changes, also added a `twos_complement` flag to give the choice whether you want to display negative numbers with a `-` sign or in the 2's complement representation.

**Review by emilk — APPROVED** [review]

**Comment by enomado** [comment]
Would be awesome to add thousands separator,  e.g. `1,000,000.0`

**Comment by Adanos020** [comment]
> Would be awesome to add thousands separator, e.g. `1,000,000.0`

@enomado Right now you can do it with `custom_formatter` I added in #1851, it's available in v0.19 :smile:
```rust
use num_format::{Locale, ToFormattedString};
ui.add(DragValue::new(&mut n).custom_formatter(|n, _| n.to_formatted_string(&Locale::en)));
```

**Pull request merged**